### PR TITLE
WIP: ENH: Add engine keyword argument to groupby.apply to leverage Numba

### LINKED ISSUE #31845: ENH: Add engine keyword argument to groupby.apply to leverage Numba
Similar to https://github.com/pandas-dev/pandas/issues/30936 and https://github.com/pandas-dev/pandas/issues/28987, we should next allow `groupby.apply` to leverage the optional Numba dependency by introducing an engine keyword that operates similarly to `rolling.apply`

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #31845
- [ ] tests added / passed
- [x] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/groupby/generic.py [hunk 1]
```diff
@@ -223,8 +223,10 @@ def _selection_name(self):
             input="series", examples=_apply_docs["series_examples"]
         )
     )
-    def apply(self, func, *args, **kwargs):
-        return super().apply(func, *args, **kwargs)
+    def apply(self, func, engine="cython", engine_kwargs=None, *args, **kwargs):
```
Comment (jreback) [inline pandas/core/groupby/generic.py]: will want to update the doc-strings as well

File: pandas/core/groupby/numba_.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
+import inspect
+import types
+
+import numpy as np
+
+from pandas.compat._optional import import_optional_dependency
+
+
+class InvalidApply(Exception):
```
Comment (jreback) [inline pandas/core/groupby/numba_.py]: why do you need a new exception?

File: pandas/core/groupby/numba_.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        function to be applied to each group and will be JITed
+    """
+    apply_function_signature = list(inspect.signature(func).parameters.keys())[:2]
+    if apply_function_signature != ["values", "index"]:
+        raise ValueError(
+            "The apply function's first 2 arguments must be 'values' and 'index'"
+        )
+
+
+def make_groupby_apply(
```
Comment (jreback) [inline pandas/core/groupby/numba_.py]: can we share pandas/core/window/numba_.py at all? maybe put common functions in pandas/core/algos/numba_.py (we have another issue about creating this and moving pandas/core/aggregation.py and algorithms.py to there

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/numba_.py (hunk 2 above)
Comment (mroeschke) [reply pandas/core/groupby/numba_.py]: Sure will want to import this from it's original location in cython. Or we can move `InvalidApply` to `pandas.errors`

File: pandas/core/groupby/numba_.py (hunk 3 above)
Comment (mroeschke) [reply pandas/core/groupby/numba_.py]: Definitely. Will work on that in another commit.

Comment (mroeschke) [reply pandas/core/groupby/numba_.py]: Definitely, will consolidate in another commit.

**Comment by mroeschke** [comment]
Closing as I am going to try to address this in another fashion

**Pull request closed without merging**